hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/client/clustering/run/strategy/MultipleSparkJobExecutionStrategy.java

```diff
@@ -420,64 +449,131 @@ writeConfig, new StoragePath(bootstrapFilePath)), partitionFields,
+
+          // instantiate other supporting cast
+          Schema readerSchema = serializableTableSchemaWithMetaFields.get();
+          Option<InternalSchema> internalSchemaOption = Option.empty();
+          if (isInternalSchemaPresent) {
+            internalSchemaOption = SerDeHelper.fromJson(internalSchemaStr);
           }
-          return readPaths;
-        })
-        .flatMap(Collection::stream)
-        .filter(path -> !path.isEmpty())
-        .map(StoragePath::new)
-        .toArray(StoragePath[]::new);
-
-    HashMap<String, String> params = new HashMap<>();
-    if (hasLogFiles) {
-      params.put("hoodie.datasource.query.type", "snapshot");
-    } else {
-      params.put("hoodie.datasource.query.type", "read_optimized");
-    }
+          Option<HoodieReaderContext> readerContextOpt = broadcastManager.retrieveFileGroupReaderContext(new StoragePath(basePath));
+          Configuration conf = broadcastManager.retrieveStorageConfig().get();
+          HoodieTableMetaClient localMetaClient = HoodieTableMetaClient.builder()
+              .setBasePath(basePath)
+              .setConf(new HadoopStorageConfiguration(conf))
+              .build();
 
-    StoragePath[] paths;
-    if (hasLogFiles) {
-      String rawFractionConfig = getWriteConfig().getString(HoodieMemoryConfig.MAX_MEMORY_FRACTION_FOR_COMPACTION);
-      String compactionFractor = rawFractionConfig != null
-          ? rawFractionConfig : HoodieMemoryConfig.DEFAULT_MR_COMPACTION_MEMORY_FRACTION;
-      params.put(HoodieMemoryConfig.MAX_MEMORY_FRACTION_FOR_COMPACTION.key(), compactionFractor);
+          // instantiate FG reader
+          HoodieFileGroupReader<T> fileGroupReader = new HoodieFileGroupReader<>(
+              readerContextOpt.get(),
+              localMetaClient.getStorage().newInstance(new StoragePath(basePath), new HadoopStorageConfiguration(conf)),
+              basePath,
+              instantTime,
+              fileSlice,
+              readerSchema,
+              readerSchema,
+              internalSchemaOption,
+              localMetaClient,
+              localMetaClient.getTableConfig().getProps(),
+              0,
+              Long.MAX_VALUE,
+              usePosition);
+          fileGroupReader.initRecordIterators();
+          // read records from the FG reader
+          HoodieFileGroupReader.HoodieFileGroupReaderIterator<InternalRow> recordIterator
+              = (HoodieFileGroupReader.HoodieFileGroupReaderIterator<InternalRow>) fileGroupReader.getClosableIterator();
+          return recordIterator;
+        }
+      }).rdd();
+
+      return HoodieUnsafeUtils.createDataFrameFromRDD(((HoodieSparkEngineContext)getEngineContext()).getSqlContext().sparkSession(),
+          internalRowRDD, sparkSchemaWithMetaFields);
+    } else {
+      boolean hasLogFiles = clusteringOps.stream().anyMatch(op -> op.getDeltaFilePaths().size() > 0);
```

Comment:
Similar to other review comments (https://github.com/apache/hudi/pull/12390/files#r1865562106), let's make two methods for if and else blocks for large blocks like this. makes the review easier (indentation stays same) and also more maintainable.

Reply:
sg. I was thinking about the same too